src/client-side-metrics/operation-metrics-collector.ts

```diff
@@ -0,0 +1,337 @@
+enum MetricsCollectorState {
+  OPERATION_NOT_STARTED,
+  OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS,
+  OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET,
+  OPERATION_STARTED_ATTEMPT_IN_PROGRESS_SOME_ROWS_RECEIVED,
+  OPERATION_COMPLETE,
+}
+
+/**
+ * A class for tracing and recording client-side metrics related to Bigtable operations.
+ */
+export class OperationMetricsCollector {
+  private state: MetricsCollectorState;
+  private operationStartTime: DateLike | null;
+  private attemptStartTime: DateLike | null;
+  private zone: string | undefined;
+  private cluster: string | undefined;
+  private tabularApiSurface: ITabularApiSurface;
+  private methodName: MethodName;
+  private attemptCount = 0;
+  private metricsHandlers: IMetricsHandler[];
+  private firstResponseLatency: number | null;
+  private serverTimeRead: boolean;
+  private serverTime: number | null;
+  private connectivityErrorCount: number;
+
+  /**
+   * @param {ITabularApiSurface} tabularApiSurface Information about the Bigtable table being accessed.
+   * @param {IMetricsHandler[]} metricsHandlers The metrics handlers used for recording metrics.
+   * @param {MethodName} methodName The name of the method being traced.
+   */
+  constructor(
+    tabularApiSurface: ITabularApiSurface,
+    metricsHandlers: IMetricsHandler[],
+    methodName: MethodName
+  ) {
+    this.state = MetricsCollectorState.OPERATION_NOT_STARTED;
+    this.zone = undefined;
+    this.cluster = undefined;
+    this.tabularApiSurface = tabularApiSurface;
+    this.methodName = methodName;
+    this.operationStartTime = null;
+    this.attemptStartTime = null;
+    this.metricsHandlers = metricsHandlers;
+    this.firstResponseLatency = null;
+    this.serverTimeRead = false;
+    this.serverTime = null;
+    this.connectivityErrorCount = 0;
+  }
+
+  private getMetricsCollectorData() {
+    return {
+      instanceId: this.tabularApiSurface.instance.id,
+      table: this.tabularApiSurface.id,
+      cluster: this.cluster,
+      zone: this.zone,
+      appProfileId: this.tabularApiSurface.bigtable.appProfileId,
+      methodName: this.methodName,
+      clientUid: this.tabularApiSurface.bigtable.clientUid,
+    };
+  }
+
+  /**
+   * Called when the operation starts. Records the start time.
+   */
+  onOperationStart() {
+    if (this.state === MetricsCollectorState.OPERATION_NOT_STARTED) {
+      this.operationStartTime = new Date();
+      this.firstResponseLatency = null;
+      this.state =
+        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS;
+    } else {
+      console.warn('Invalid state transition');
+    }
+  }
+
+  /**
+   * Called when an attempt (e.g., an RPC attempt) completes. Records attempt latencies.
+   * @param {string} projectId The id of the project.
+   * @param {OnAttemptCompleteInfo} info Information about the completed attempt.
+   */
+  onAttemptComplete(projectId: string, info: OnAttemptCompleteInfo) {
```

Comment:
why does a StreamingState need to be passed here? and project id? shouldn't they be known in advance?

It seems to me that the only new information here would be the status code

Reply:
We could definitely pass in the streamingOperation parameter when we create the metrics collector instead so we don't need to pass it in here. projectId is a different story, because the projectId property on the client is only populated by calling auth just before the grpc call is made so it's not available when the metrics collector gets created.